transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,459 @@
+            rx,
+            port: self.port,
+            max_datagrams: self.max_datagrams,
+        };
+        (channel, tx)
+    }
+}
+
+type QuinnEndpointConfig<S> = quinn_proto::generic::EndpointConfig<S>;
+type QuinnEndpoint<S> = quinn_proto::generic::Endpoint<S>;
+
+pub struct EndpointConfig<C: Crypto> {
+    socket: UdpSocket,
+    endpoint: QuinnEndpoint<C::Session>,
+    port: u16,
+    crypto_config: Arc<CryptoConfig<C>>,
+    capabilities: UdpCapabilities,
+}
+
+impl<C: Crypto> EndpointConfig<C> {
+    pub fn new(mut config: QuicConfig<C>, addr: SocketAddr) -> Result<Self, QuicError> {
+        config.transport.max_concurrent_uni_streams(0)?;
+        config.transport.datagram_receive_buffer_size(None);
+        let transport = Arc::new(config.transport);
+
+        let crypto_config = Arc::new(CryptoConfig {
+            keypair: config.keypair,
+            psk: config.psk,
+            keylogger: config.keylogger,
+            transport: transport.clone(),
+        });
+
+        let mut server_config = ServerConfig::<C::Session>::default();
+        server_config.transport = transport;
+        server_config.crypto = C::new_server_config(&crypto_config);
+
+        let mut endpoint_config = QuinnEndpointConfig::default();
+        endpoint_config
+            .supported_versions(C::supported_quic_versions(), C::default_quic_version())?;
+
+        let socket = UdpSocket::bind(addr)?;
+        let port = socket.local_addr()?.port();
+        let endpoint = quinn_proto::generic::Endpoint::new(
+            Arc::new(endpoint_config),
+            Some(Arc::new(server_config)),
+        );
+        let capabilities = UdpSocket::capabilities()?;
+        Ok(Self {
+            socket,
+            endpoint,
+            port,
+            crypto_config,
+            capabilities,
+        })
+    }
+
+    pub fn spawn(self) -> TransportChannel<C>
+    where
+        <C::Session as Session>::ClientConfig: Send + Unpin,
+        <C::Session as Session>::HeaderKey: Unpin,
+        <C::Session as Session>::PacketKey: Unpin,
+    {
+        let (tx1, rx1) = mpsc::unbounded();
+        let (tx2, rx2) = mpsc::channel(1);
+        let transport = TransportChannel {
+            tx: tx1,
+            rx: rx2,
+            port: self.port,
+            ty: self.socket.socket_type(),
+        };
+        let endpoint = EndpointChannel {
+            tx: tx2,
+            rx: rx1,
+            port: self.port,
+            max_datagrams: self.capabilities.max_gso_segments,
+            connection_tx: transport.tx.clone(),
+        };
+        async_global_executor::spawn(Endpoint::new(endpoint, self)).detach();
+        transport
+    }
+}
+
+struct Endpoint<C: Crypto> {
+    channel: EndpointChannel<C>,
+    endpoint: QuinnEndpoint<C::Session>,
+    socket: UdpSocket,
+    crypto_config: Arc<CryptoConfig<C>>,
+    connections: FnvHashMap<ConnectionHandle, mpsc::Sender<ConnectionEvent>>,
+    outgoing: VecDeque<udp_socket::Transmit>,
```

Comment:
Is there a proof that this cannot get too large?  If it does get over a certain size, would it be better to drop packets, as e.g. the kernel network stack does?

Reply:
These are outgoing, so you want to drop packets to prevent libp2p dos'ing itself? As far as I know this pattern is used throughout the codebase

Reply:
Is it possible for a peer to create a situation where libp2p tries to send a lot of data, but then is not actually able to do so?  That said, I do not believe this should block merging if it is not a regression compared to the other parts of the codebase.

Reply:
The first packet requires a large payload before the server will answer. This restriction is lifted after the ip address was validated to prevent amplification attacks. Later on I guess that would depend on the application protocol. If you are concerned about this it might be a good idea to bandwidth limit ip's using the bandwidth wrapper.

Reply:
Marking this as resolved

Reply:
How is this different to an unbounded channel? How is back-pressure to the upper layer enforced?

Reply:
queues are used throughout the codebase, don't know why we're making a fuss about it here